### PR TITLE
NEXUS-21322: Component original path

### PR DESCRIPTION
**Pull request by chernikovmaksym**
https://issues.sonatype.org/browse/NEXUS-21322

### CODE REVIEW

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RProxyFacetImpl.java [hunk 1]
```diff
@@ -124,76 +118,55 @@ public void setCacheInfo(final Content content, final CacheInfo cacheInfo) {
     tx.saveAsset(asset);
   }
 
-  private Content putArchive(final String path, final String filename, final Content content) throws IOException {
+  private Content putArchive(final String path, final Content content) throws IOException {
+    checkNotNull(path);
+    checkNotNull(content);
```
Comment (sergeyonofreychuck) [inline nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RProxyFacetImpl.java]: do we need to check for null private and protected methods? 
There is no @nullable modifiers on any parameters

**Comment by sergeyonofreychuck** [comment]
Split the refactoring and the task implementation to separate PRs.

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java [hunk 2]
```diff
@@ -80,7 +80,7 @@ public Asset findOrCreateAsset(final StorageTx tx,
       for (Entry<String, String> attribute : attributes.entrySet()) {
         asset.formatAttributes().set(attribute.getKey(), attribute.getValue());
       }
-      asset.formatAttributes().set(P_ASSET_KIND, getAssetKind(path));
+      asset.formatAttributes().set(P_ASSET_KIND, determineAssetKind(path).name());
```
Comment (sergeyonofreychuck) [inline nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java]: the old name I like more than the new

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java [hunk 3]
```diff
@@ -95,20 +95,17 @@ static String cutFilenameFromPath(final String path) {
...
-      assetKind = PACKAGES.name();
+      return PACKAGES;
     }
     else if (Pattern.compile(PATTERN_METADATA_RDS).matcher(assetPath).matches()) {
-      assetKind = RDS_METADATA.name();
+      return RDS_METADATA;
+    }
+    else {
+      return ARCHIVE;
     }
```
Comment (sergeyonofreychuck) [inline nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java]: in NXRM mostly used the format like in the code before your changes. multiple returns increases complexity of a function

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RProxyFacetImpl.java (hunk 1 above)
Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RProxyFacetImpl.java]: done in 7a2327f930a2b894b69a6e3faae6338d3eadae6e

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java (hunk 2 above)
Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java]: done in 7a2327f

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java (hunk 3 above)
Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java]: done in 7a2327f

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java [hunk 4]
```diff
@@ -70,14 +71,16 @@ public Asset findOrCreateAsset(final StorageTx tx,
...
       asset.name(path);
-      for (Map.Entry attribute : attributes.entrySet()) {
-        asset.formatAttributes().set(attribute.getKey().toString(), attribute.getValue());
+
+      // TODO: Make this a bit more robust (could be problematic if keys are removed in later versions, or if keys clash)
+      for (Entry<String, String> attribute : attributes.entrySet()) {
+        asset.formatAttributes().set(attribute.getKey(), attribute.getValue());
       }
-      asset.formatAttributes().set(P_ASSET_KIND, getAssetKind(path));
+      asset.formatAttributes().set(P_ASSET_KIND, getAssetKind(path).name());
```
Comment (doddi) [inline nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java]: Do we need to consider existing R assets? I think we should possibly consider an upgrade process

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java [hunk 5]
```diff
@@ -75,17 +82,30 @@ static String packagesGzPath(final String path) {
...
+  static String cutFilenameFromPath(final String path) {
+    final int pathEndCharIndex = StringUtils.stripEnd(path, "/").lastIndexOf('/');
+    return path.substring(0, pathEndCharIndex);
+  }
+
+  /**
+   * Determines asset kind by it's path
+   */
+  static AssetKind getAssetKind(final String path) {
+    AssetKind assetKind = ARCHIVE;
```
Comment (doddi) [inline nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java]: Question: Is it safe to assume `ARCHIVE`? is it possible for an invalid path to be here?

Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java]: Yes, cause this method looks only on stored assets. So if it's not any kind of metadata it's definitely archive.

Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPathUtils.java]: BTW, I just moved this method without big modification

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java (hunk 4 above)
Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java]: You're talking about assetkind in attributes, yes? If so, master version doesn't support any kind of `.rds` metadata and this assetkind is not used on previous versions. All others are compatible with master and no need to make upgrade process.

**Comment by chernikovmaksym** [comment]
> I know we have not catered explicitly for existing community format but given the huge win/positive feedback we got, I think we should strive to accommodate existing R users. As a minimum add an upgrade step.

I consider all updates to be compatible with previous versions (it was tested some PRs ago). And it's going to be fully tested again before release. If it needs any compatibility upgrades it will be discussed and done.

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java (hunk 4 above)
Comment (chernikovmaksym) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RFacetImpl.java]: No, method used to return string representation before, I just decided it was a bit confusing

**Pull request merged**